### ISSUE TITLE
Reconnection of websocket client

### ISSUE THREAD

**Issue by jonaskello** [cloudamqp/amqp-client.js#25]
I'm trying to implement reconnection logic for the `AMQPWebSocketClient`.

I have two scenarios I'm trying to cover:

1. The server is down when the client starts up. Then the server comes up and the client should connect automatically.
2. The server is up when the client starts up and the client connects. Later the server goes down for a while and then comes back up. The client should reconnect automatcially when the server is back up.

I have an [example here](https://github.com/jonaskello/amqp-example-ws-ts-react-webpack) to try out these scenarios.

Scenario (1) works but I'm having some trouble to make scenario (2) work.

Basically I do this to simulate scenario (2):

1. Start rabbit and relay: `docker compose up -d`.
2. Start the client: `yarn start`
3. Open web browser. The client connects and the Send button works.
4. Check id of the relay container and stop it so simulate server going down:

```
$ docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED         STATUS         PORTS                                                                                                         NAMES
1b607d716888   rabbitmq:3.9.13-management      "docker-entrypoint.s…"   4 minutes ago   Up 4 minutes   4369/tcp, 5671/tcp, 0.0.0.0:5672->5672/tcp, 15671/tcp, 15691-15692/tcp, 25672/tcp, 0.0.0.0:15672->15672/tcp   amqp-example-ws-ts-react-webpack-rabbitmq-1
11c9052798d6   cloudamqp/websocket-tcp-relay   "/usr/local/bin/webs…"   4 minutes ago   Up 4 minutes   0.0.0.0:15670->15670/tcp                                                                                     amqp-example-ws-ts-react-webpack-relay-1
$ docker stop 11c9052798d6
11c9052798d6
```

5. Ideally the client would just start reconnecting automatically at this point but I have not found an API in the library to listen to the socket state. Is there such an API? If not I think that would be a nice addition :-).

6. Click the Send button to try to send new message. This code has an exception handler that calls the connection logic. However the problem is that even though all code in the package and in the example seem to have correct exception handling it is not called. Instead the web browser writes this in the console:

```
WebSocket is already in CLOSING or CLOSED state.
(anonymous)	                              @	amqp-websocket-client.js:30
send	                                      @	amqp-websocket-client.js:27
basicPublish                                  @	amqp-channel.js:345
(anonymous)                                   @	index.tsx:51
(anonymous)                                   @	tracing.js:7
__webpack_modules__../src/index.tsx.__awaiter @ tracing.js:7
publishMessage	                              @	index.tsx:48
onClick	                                      @	index.tsx:64
callCallback	                              @	react-dom.development.js:394
```

Even though the console message looks like an exception it is not caught by the handlers. I'm puzzled by this. Looking at the line that is the root in the stack trace is looks like this:

```
  send(bytes) {
        return new Promise((resolve, reject) => {
            if (this.socket) {
                try {
                    this.socket.send(bytes); // this line
                    resolve();
                }
                catch (err) {
                    reject(err);
                }
            }
            else {
                reject("Socket not connected");
            }
        });
    }
```

Obviously the socket state is "CLOSING or CLOSED" as the message says. Reading in the [MDN docs for websocket.send()](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send) I found this:

> If you call send() when the connection is in the CLOSING or CLOSED states, the browser will silently discard the data.

I'm not really sure what to make of this. Perhaps one possible explanation could be that a "real" exception is not thrown, instead just an error with stacktrace is logged an the browser silently discard the data.

The code in my example basically implements the code from the readme but in a typescript/react/webpack setting. However if I run the exact example code from the readme using just javascript with `<script type="module">` it does seem to actually throw an exception that can be caught.

Perhaps I'm doing something wrong in my example that causes the exception not to propagate but I cannot see what it would be.

**Comment by jonaskello** [comment]
Actually the `<script type=module>` example also gives the same message when I tried it again now. 

I put an example that is exactly like the readme [here](https://github.com/jonaskello/amqp-example-ws-js). It also gives the `WebSocket is already in CLOSING or CLOSED state.` message when following the same steps as above.

**Comment by carlhoerberg** [comment]
`AMQPClient#onerror` is available Included in v2.1.0
